modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxResourceCleanupManager.java

```diff
@@ -0,0 +1,117 @@
+ */
+public class TxResourceCleanupManager implements IgniteComponent {
+    /** The logger. */
+    private static final IgniteLogger LOG = Loggers.forClass(TxResourceCleanupManager.class);
+
+    private static final int RESOURCE_CLEANUP_EXECUTOR_SIZE = 1;
+
+    /** System property name. */
+    public static final String RESOURCE_CLEANUP_INTERVAL_MILLISECONDS_PROPERTY = "RESOURCE_CLEANUP_INTERVAL_MILLISECONDS";
+
+    private final int resourceCleanupIntervalMilliseconds = IgniteSystemProperties
+            .getInteger(RESOURCE_CLEANUP_INTERVAL_MILLISECONDS_PROPERTY, 30_000);
+
+    private final ScheduledExecutorService resourceCleanupExecutor;
+
+    private final RemotelyTriggeredResourceRegistry resourceRegistry;
+
+    private final ClusterNodeResolver clusterNodeResolver;
+
+    /**
+     * Constructor.
+     *
+     * @param nodeName Name of the Ignite node.
+     * @param resourceRegistry Remote triggered resource registry.
+     * @param clusterNodeResolver Cluster node resolver.
+     */
+    public TxResourceCleanupManager(
+            String nodeName,
+            RemotelyTriggeredResourceRegistry resourceRegistry,
+            ClusterNodeResolver clusterNodeResolver
+    ) {
+        resourceCleanupExecutor = Executors.newScheduledThreadPool(
+                RESOURCE_CLEANUP_EXECUTOR_SIZE,
+                NamedThreadFactory.create(nodeName, "resource-cleanup-executor", LOG)
+        );
+        this.resourceRegistry = resourceRegistry;
+        this.clusterNodeResolver = clusterNodeResolver;
+    }
+
+    @Override
+    public CompletableFuture<Void> start() {
+        resourceCleanupExecutor.scheduleAtFixedRate(
+                this::cleanupOrphanTxResources,
+                0,
+                resourceCleanupIntervalMilliseconds,
+                TimeUnit.MILLISECONDS
+        );
+
+        return nullCompletedFuture();
+    }
+
+    @Override
+    public void stop() {
+        shutdownAndAwaitTermination(resourceCleanupExecutor, 10, TimeUnit.SECONDS);
+    }
+
+    private void cleanupOrphanTxResources() {
+        Map<FullyQualifiedResourceId, RemotelyTriggeredResource> cursorInfos = resourceRegistry.resources();
+
+        try {
+            for (Map.Entry<FullyQualifiedResourceId, RemotelyTriggeredResource> cursorInfoEntry : cursorInfos.entrySet()) {
+                RemotelyTriggeredResource cursorInfo = cursorInfoEntry.getValue();
+
+                if (clusterNodeResolver.getById(cursorInfo.remoteHostId()) == null) {
+                    try {
+                        resourceRegistry.close(cursorInfoEntry.getKey());
+                    } catch (Exception e) {
+                        LOG.warn("Error occured during the orphan cursor closing [txCoordinatorId={}]", e, cursorInfo.remoteHostId());
+                    }
+                }
+            }
+        } catch (Throwable err) {
+            LOG.error("Error occured during the orphan cursors closing.", err);
+
+            if (err instanceof Error) {
```

Comment:
The point is to log everything that can happen, otherwise we will just not see all errors. And we assume that assertion error is something that should never happen, why should it be handled differently?